### ISSUE TITLE
Problem with in logic for determining ${ARCH_FLAG} and ${TUNE_FLAG}

### ISSUE THREAD

**Issue by graysky2** [VectorCamp/vectorscan#153]
If I am building for OpenWRT, the logic in `CMakeList.txt` for determining `${ARCH_FLAG}` and `${TUNE_FLAG}` seems to define invalid choices for both variables.  I am not expert enough to suggest an alternative.  You can see below the errors:
```
% make package/vectorscan/compile package/index V=s
make[2]: Entering directory '/scratch/union/scripts/config'
make[2]: 'conf' is up to date.
make[2]: Leaving directory '/scratch/union/scripts/config'
make[1]: Entering directory '/scratch/union'
make[2]: Entering directory '/scratch/union/feeds/packages/libs/boost'
make[2]: Entering directory '/scratch/union/feeds/packages/libs/libffi'
make[2]: Entering directory '/scratch/union/package/utils/bzip2'
make[2]: Entering directory '/scratch/union/feeds/packages/devel/ragel'
make[2]: Entering directory '/scratch/union/package/libs/toolchain'
make[2]: Leaving directory '/scratch/union/feeds/packages/devel/ragel'
time: package/feeds/packages/ragel/host-compile#0.07#0.02#0.09
Selected Boost API aapcs for architecture aarch64 and cpu type "cortex-a72" 
echo "libc" >> /scratch/union/staging_dir/target-aarch64_cortex-a72_musl/pkginfo/toolchain.default.install
echo "libgcc" >> /scratch/union/staging_dir/target-aarch64_cortex-a72_musl/pkginfo/toolchain.default.install
echo "libpthread" >> /scratch/union/staging_dir/target-aarch64_cortex-a72_musl/pkginfo/toolchain.default.install
echo "librt" >> /scratch/union/staging_dir/target-aarch64_cortex-a72_musl/pkginfo/toolchain.default.install
make[2]: Leaving directory '/scratch/union/package/libs/toolchain'
time: package/libs/toolchain/compile#0.14#0.02#0.16
make[2]: Leaving directory '/scratch/union/feeds/packages/libs/libffi'
time: package/feeds/packages/libffi/host-compile#0.15#0.03#0.16
make[2]: Leaving directory '/scratch/union/package/utils/bzip2'
time: package/utils/bzip2/host-compile#0.18#0.02#0.20
make[2]: Entering directory '/scratch/union/feeds/packages/lang/python/python3'
make[2]: Leaving directory '/scratch/union/feeds/packages/lang/python/python3'
time: package/feeds/packages/python3/host-compile#0.20#0.04#0.23
make[2]: Leaving directory '/scratch/union/feeds/packages/libs/boost'
time: package/feeds/packages/boost/host-compile#0.68#0.27#0.77
make[2]: Entering directory '/scratch/union/feeds/packages/utils/vectorscan'
rm -f /scratch/union/build_dir/target-aarch64_cortex-a72_musl/vectorscan-vectorscan-5.4.9/.configured_*
rm -f /scratch/union/staging_dir/target-aarch64_cortex-a72_musl/stamp/.vectorscan_installed
mkdir -p /scratch/union/build_dir/target-aarch64_cortex-a72_musl/vectorscan-vectorscan-5.4.9
(cd /scratch/union/build_dir/target-aarch64_cortex-a72_musl/vectorscan-vectorscan-5.4.9; CFLAGS="-Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -fmacro-prefix-map=/scratch/union/build_dir/target-aarch64_cortex-a72_musl/vectorscan-vectorscan-5.4.9=vectorscan-vectorscan-5.4.9 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro " CXXFLAGS="-Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -fmacro-prefix-map=/scratch/union/build_dir/target-aarch64_cortex-a72_musl/vectorscan-vectorscan-5.4.9=vectorscan-vectorscan-5.4.9 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro " LDFLAGS="-L/scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/usr/lib -L/scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/lib -znow -zrelro " cmake -DCMAKE_SYSTEM_NAME=Linux -DCMAKE_SYSTEM_VERSION=1 -DCMAKE_SYSTEM_PROCESSOR=aarch64 -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_FLAGS_RELEASE="-DNDEBUG" -DCMAKE_CXX_FLAGS_RELEASE="-DNDEBUG" -DCMAKE_C_COMPILER_LAUNCHER="" -DCMAKE_C_COMPILER="/scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/bin/aarch64-openwrt-linux-musl-gcc" -DCMAKE_CXX_COMPILER_LAUNCHER="" -DCMAKE_CXX_COMPILER="/scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/bin/aarch64-openwrt-linux-musl-g++" -DCMAKE_ASM_COMPILER_LAUNCHER="" -DCMAKE_ASM_COMPILER="/scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/bin/aarch64-openwrt-linux-musl-gcc" -DCMAKE_EXE_LINKER_FLAGS:STRING="-L/scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/usr/lib -L/scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/lib -znow -zrelro" -DCMAKE_MODULE_LINKER_FLAGS:STRING="-L/scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/usr/lib -L/scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/lib -znow -zrelro -Wl,-Bsymbolic-functions" -DCMAKE_SHARED_LINKER_FLAGS:STRING="-L/scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/usr/lib -L/scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/lib -znow -zrelro -Wl,-Bsymbolic-functions" -DCMAKE_AR="/scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/bin/aarch64-openwrt-linux-musl-gcc-ar" -DCMAKE_NM="/scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/bin/aarch64-openwrt-linux-musl-gcc-nm" -DCMAKE_RANLIB="/scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/bin/aarch64-openwrt-linux-musl-gcc-ranlib" -DCMAKE_FIND_ROOT_PATH="/scratch/union/staging_dir/target-aarch64_cortex-a72_musl/usr;/scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl" -DCMAKE_FIND_ROOT_PATH_MODE_PROGRAM=BOTH -DCMAKE_FIND_ROOT_PATH_MODE_LIBRARY=ONLY -DCMAKE_FIND_ROOT_PATH_MODE_INCLUDE=ONLY -DCMAKE_STRIP=: -DCMAKE_INSTALL_PREFIX=/usr -DDL_LIBRARY=/scratch/union/staging_dir/target-aarch64_cortex-a72_musl -DCMAKE_PREFIX_PATH=/scratch/union/staging_dir/target-aarch64_cortex-a72_musl -DCMAKE_SKIP_RPATH=TRUE -DCMAKE_EXPORT_PACKAGE_REGISTRY=FALSE -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=TRUE -DCMAKE_FIND_USE_PACKAGE_REGISTRY=FALSE -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=TRUE -DCMAKE_FIND_USE_SYSTEM_PACKAGE_REGISTRY=FALSE -DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=TRUE -DCMAKE_GENERATOR="Ninja" -DCMAKE_INSTALL_LIBDIR=lib -DBUILD_SHARED_LIBS=ON -Wno-dev /scratch/union/build_dir/target-aarch64_cortex-a72_musl/vectorscan-vectorscan-5.4.9 )
-- The C compiler identification is GNU 12.3.0
-- The CXX compiler identification is GNU 12.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/bin/aarch64-openwrt-linux-musl-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /scratch/union/staging_dir/toolchain-aarch64_cortex-a72_gcc-12.3.0_musl/bin/aarch64-openwrt-linux-musl-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test ARCH_X86_64
-- Performing Test ARCH_X86_64 - Failed
-- Performing Test ARCH_IA32
-- Performing Test ARCH_IA32 - Failed
-- Performing Test ARCH_AARCH64
-- Performing Test ARCH_AARCH64 - Success
-- Performing Test ARCH_ARM32
-- Performing Test ARCH_ARM32 - Failed
-- Performing Test ARCH_PPC64EL
-- Performing Test ARCH_PPC64EL - Failed
-- Build type RELEASE
-- using release build
-- Boost version: 1.82.0
-- Found Python: /scratch/union/staging_dir/hostpkg/bin/python3.11 (found version "3.11.3") found components: Interpreter 
-- Build date: 
-- Building shared libraries
-- gcc version 12.3.0
-- ARCH_FLAG 'march' '', TUNE_FLAG 'mtune' '' 
cc1: error: unknown value 'native' for '-march'
cc1: note: valid arguments are: armv8-a armv8.1-a armv8.2-a armv8.3-a armv8.4-a armv8.5-a armv8.6-a armv8.7-a armv8.8-a armv8-r armv9-a
cc1: error: unknown value 'native' for '-mtune'
cc1: note: valid arguments are: cortex-a34 cortex-a35 cortex-a53 cortex-a57 cortex-a72 cortex-a73 thunderx thunderxt88p1 thunderxt88 octeontx octeontx81 octeontx83 thunderxt81 thunderxt83 ampere1 ampere1a emag xgene1 falkor qdf24xx exynos-m1 phecda thunderx2t99p1 vulcan thunderx2t99 cortex-a55 cortex-a75 cortex-a76 cortex-a76ae cortex-a77 cortex-a78 cortex-a78ae cortex-a78c cortex-a65 cortex-a65ae cortex-x1 ares neoverse-n1 neoverse-e1 octeontx2 octeontx2t98 octeontx2t96 octeontx2t93 octeontx2f95 octeontx2f95n octeontx2f95mm a64fx tsv110 thunderx3t110 zeus neoverse-v1 neoverse-512tvb saphira cortex-a57.cortex-a53 cortex-a72.cortex-a53 cortex-a73.cortex-a35 cortex-a73.cortex-a53 cortex-a75.cortex-a55 cortex-a76.cortex-a55 cortex-r82 cortex-a510 cortex-a710 cortex-x2 neoverse-n2 demeter neoverse-v2 generic
-- ARCH_FLAG 'march' 'native', TUNE_FLAG 'mtune' 'native' 
CMake Warning at CMakeLists.txt:207 (message):
  Something went wrong determining gcc tune: -mtune=native not valid, falling
  back to -mtune=native
```

**Comment by graysky2** [comment]
I am not skilled enough to formulate a PR.  Let's see what the others I mentioned have to say.

**Comment by jefferyto** [comment]
We set `CMAKE_SYSTEM_NAME` to `Linux` ([here](https://github.com/openwrt/openwrt/blob/bf8187d5dc4d4bbb23770955744bca1787f32ac0/include/cmake.mk#L92)).

I'm not sure if there is any real issue though as I'm fairly certain the [instructions for cross-compiling](https://github.com/VectorCamp/vectorscan#cross-compiling-for-aarch64) hasn't been followed/adapted yet.

**Comment by graysky2** [comment]
> I'm not sure if there is any real issue though as I'm fairly certain the [instructions for cross-compiling](https://github.com/VectorCamp/vectorscan#cross-compiling-for-aarch64) hasn't been followed/adapted yet.

I am not totally sure how to do this in our `Makefile`; grateful for any suggestions.

**Comment by markos** [comment]
The instructions for cross-compiling are out of date and probably inaccurate now. The only source of truth right now is the actual CMakeFiles.txt file.

**Comment by sirrkitt** [comment]
Trying to build this in a container running Wolfi OS and it's giving me the same issue:

`g++: error: missing argument to '-march='`
`g++: error: missing argument to '-march='`

Trying to build with GCC 13.2.0 and glibc 2.38 and running on arm64.

If I switch to using a Debian base it works, though. I usually set CFLAGS to include:
`-march=armv8-a+crc+crypto+fp16+rcpc -mtune=generic`

I get the same error regardless of whether or not I set -march in CFLAGS

**Comment by steev** [comment]
We are seeing similar build failures in Kali's CI for pyperscan which uses vectorscan under the hood.

With gcc-13, we see
```
[hyperscan-sys 0.2.2] -- ARCH_FLAG 'march' '', TUNE_FLAG 'mtune' '' 
[hyperscan-sys 0.2.2] -- ARCH_FLAG 'march' '', TUNE_FLAG 'mtune' 'emag' 
```
And forcing the build system back to gcc-12, we see
```
[hyperscan-sys 0.2.2] -- ARCH_FLAG 'march' '', TUNE_FLAG 'mtune' ''
[hyperscan-sys 0.2.2] -- ARCH_FLAG 'march' 'armv8.2-a+crypto+fp16+rcpc+dotprod+flagm', TUNE_FLAG 'mtune' 'generic'
[hyperscan-sys 0.2.2] -- gcc will tune for armv8.2-a+crypto+fp16+rcpc+dotprod+flagm, generic
```

https://gitlab.com/kalilinux/packages/pyperscan/-/jobs/5013652105 is a link to our failed build in CI, which is a docker container running on LXC, but I also see the same locally running on a Qualcomm SC8280XP based Thinkpad X13s, and on a ROCK-5B which has a RK3588.

**Comment by markos** [comment]
Could you please try current develop branch, we believe the issue may be fixed in #181.

**Comment by gotkube** [comment]
> Could you please try current develop branch, we believe the issue may be fixed in #181.

FYI, I successfully built the develop branch yesterday on a RPI3 with gcc 13.2.0 & glibc 2.37.
I was having similar problems mentioned in this thread and this seems to have resolved them. Thanks! :)
